ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -162,14 +165,30 @@ runGddAsync loEUpdater varLoEFrag chainDb getTrigger =
       triggerChainSelectionAsync chainDb
       spin newTrigger
 
+-- | Classify the latest known header in relation to the forecast horizon,
+-- and provide its slot number.
+--
+data LatestSlot =
+  -- | The candidate fragment is empty and the TVar containing the latest
+  -- slot ChainSync has seen does not contain an entry for the current peer.
+  NoLatestSlot
+  |
+  -- | The candidate fragment head is the latest known header.
+  LatestSlotCandidate SlotNo
+  |
+  -- | ChainSync has seen a header after the candidate fragment head, most
+  -- likely because it is beyond the forecast horizon.
+  LatestSlotForecast SlotNo
```

Comment:
How about we change `LatestSlotForecast` to `LatestSlotAfterSGen`? I don't think we care about the forecast horizon as much as we care of the header being beyond the genesis window.

Reply:
the point of `Forecast` is that the block is not on the candidate fragment, so we have to count differently. Even with `LatestSlotCandidate` the last block can be after `sgen`

Reply:
This is only relevant for `offersMoreThanK`, where a candidate fragment with `k` blocks can be considered to meet the condition because the last slot implies that there is a header outside of the candidate fragment.

The constructor is mostly for readability I guess? Though this conversation suggests that it's not intuitive enough

Reply:
maybe it should be

```haskell
| LatestSlotBeforeSgen
| LatestSlotAfterSgen { forecast :: Bool }
```

Would you find that more accessible?

Reply:
> This is only relevant for `offersMoreThanK`

The meaning of `LatestSlotForecast` affects the computation of the upper bound as well.

Reply:
> Would you find that more accessible?

I still don't understand why we would care of the last header being after the forecast horizon. See my comment in `offersMoreThanK`.

Reply:
I eliminated the distinction between Candidate and Forecast as it wasn't needed for `offersMoreThanK`. This is the current form:
```haskell
data LatestSlot = NoLatestSlot | LatestSlot SlotNo
```
In the case of the computation of the upper bound, it isn't dangerous to use the slots of not-yet-validated headers after the genesis window, as it allows to prune adversarial nodes early.

Reply:
> The meaning of LatestSlotForecast affects the computation of the upper bound as well.

btw that's just a simplification, because we know that the difference will always be `<= 0`